lib/cartopy/mpl/gridliner.py

```diff
@@ -327,58 +338,257 @@ def _draw_gridliner(self, nx=None, ny=None, background_patch=None):
+        map_boundary = sgeom.Polygon(map_boundary_vertices)
+
+        self._labels = []
+        for lonlat, lines, line_ticks, formatter, label_style in (
+                ('lon', lon_lines, lon_ticks,
+                 self.xformatter, self.xlabel_style),
+                ('lat', lat_lines, lat_ticks,
+                 self.yformatter, self.ylabel_style)):
+
+            formatter.set_locs(line_ticks)
+
+            for line, tick_value in zip(lines, line_ticks):
+
+                # Intersection of line with map boundary
+                line = np.array(line)
+                line = self.axes.projection.transform_points(
+                        crs, line[:, 0], line[:, 1])[:, :2]
+                infs = np.isinf(line)
+                if infs.any():
+                    if infs.all():
+                        continue
+                    line = line.compress(~infs.any(axis=1), axis=0)
+                line = sgeom.LineString(line)
+                if line.intersects(map_boundary):
+                    intersection = line.intersection(map_boundary)
+                    del line
+                    if intersection.is_empty:
+                        continue
+                    if isinstance(intersection, sgeom.MultiPoint):
+                        if len(intersection) < 2:
+                            continue
+                        tails = [[(pt.x, pt.y) for pt in intersection[:2]]]
+                        heads = [[(pt.x, pt.y)
+                                  for pt in intersection[-1:-3:-1]]]
+                    elif isinstance(intersection, (sgeom.LineString,
+                                                   sgeom.MultiLineString)):
+                        if isinstance(intersection, sgeom.LineString):
+                            intersection = [intersection]
+                        elif len(intersection) > 4:
+                            # gridline and map boundary are parallel
+                            # and they intersect themselve too musc
+                            # it results in a multiline string
+                            # that must be converted to a single linestring
+                            x, y = intersection[0].coords.xy
+                            x += intersection[-1].coords.xy[0]
+                            y += intersection[-1].coords.xy[1]
+                            xy = np.array((x, y))
+                            intersection = [sgeom.LineString(xy.T)]
+                        tails = []
+                        heads = []
+                        for inter in intersection:
+                            if len(inter.coords) < 2:
+                                continue
+                            tails.append(inter.coords[:2])
+                            heads.append(inter.coords[-1:-3:-1])
+                        if not tails:
+                            continue
+                    else:
+                        # TODO: we should handle GeometryCollection
```

Comment:
Can you please put a warning or error in here (maybe a `NotImplementedError`, although a warning may be more user-friendly) explaining to the user that lots of processing isn't being done here and why.

Reply:
I handled the GeometryCollection base.
I'm not sure of other cases may occurs, so I added a warning if one of them happens.
However, a warning may not be appropriate if the user cannot do anything, and raising an error is probably too much, since it will simply not plot some labels, which occurs often for other reasons.